src/datahike/index/persistent_set.cljc

```diff
@@ -31,54 +24,103 @@
-    (set/from-sorted-array (index-type->cmp index-type) arr)))
-
-(defn -remove [set datom index-type]
-  (set/disj set datom (index-type->cmp-quick index-type)))
-
-(defn -insert [set datom index-type]
-  (-> (or (when-let [old (first (-slice set
-                                        (dd/datom (.-e datom) (.-a datom) (.-v datom) tx0)
-                                        (dd/datom (.-e datom) (.-a datom) (.-v datom) txmax)))]
-            (-remove set old index-type))
-          set)
-      (set/conj datom (index-type->cmp-quick index-type))))
-
-(defn -temporal-insert [set datom index-type]
-  (set/conj set datom (index-type->cmp-quick index-type)))
-
-(defn -upsert [set datom index-type]
-  (-> (or (when-let [old (first (-slice set
-                                        (dd/datom (.-e datom) (.-a datom) nil tx0)
-                                        (dd/datom (.-e datom) (.-a datom) nil txmax)))]
-            (-remove set old index-type))
-          set)
-      (set/conj datom (index-type->cmp-quick index-type))))
-
-(defn -temporal-upsert [set datom index-type]
-  (-> (or (when-let [old (first (-slice set
-                                        (dd/datom (.-e datom) (.-a datom) nil tx0)
-                                        (dd/datom (.-e datom) (.-a datom) nil txmax)))]
-            (set/conj set (dd/datom (.-e old) (.-a old) (.-v old) (.-tx old) false)
-                      (index-type->cmp-quick index-type)))
-          set)
-      (set/conj datom (index-type->cmp-quick index-type))))
+                (arrays/into-array)))]
+    (arrays/asort arr (index-type->cmp-quick index-type))
+    (with-meta (set/from-sorted-array (index-type->cmp index-type) arr (get-loader store))
+      {:index-type index-type})))
+
+(defmethod di/add-konserve-handlers :datahike.index/persistent-set [_index-name store]
+  (assoc store :serializers {:FressianSerializer (fressian-serializer
+                                                  {"datahike.index.PersistentSortedSet" (reify ReadHandler
+                                                                                          (read [_ reader _tag _component-count]
+                                                                                            (let [{:keys [index-type root-node]} (.readObject reader)
+                                                                                                  loader (get-loader store)
+                                                                                                  pset (with-meta (set/sorted-set-by (index-type->cmp index-type) loader)
+                                                                                                         {:index-type index-type})]
+                                                                                              (set! (._root pset)
+                                                                                                    (set/map->node loader root-node))
```

Comment:
It would be better to provide serializers for the Node/Leaf class of the persistent sorted set itself, then we would not have to (de)serialize them in the loader, but can instead do it transparently in the storage layer.

Reply:
This is not critical though, in particular if we do manual migrations for now anyway.